File: phases/07-transformers-deep-dive/04-positional-encoding/code/main.jl
# Positional encoding in Julia. Sinusoidal absolute positions, rotary
# positional embedding (RoPE), and ALiBi bias matrix. Verifies that
# RoPE dot products depend only on relative distance. Stdlib only. Sources:
#   https://arxiv.org/abs/2104.09864
#   https://arxiv.org/abs/2108.12409
#   https://docs.julialang.org/en/v1/manual/mathematical-operations/

using Random
using Printf


function sinusoidal_pe(n::Int, d::Int; base::Float64=10000.0)::Matrix{Float64}
    n > 0 || throw(ArgumentError("n must be > 0"))
    d > 0 || throw(ArgumentError("d must be > 0"))
    iseven(d) || throw(ArgumentError("d must be even for sinusoidal sin/cos pairs"))
    pe = zeros(n, d)
    for pos in 0:(n - 1)
        for i in 0:(d ÷ 2 - 1)
            theta = pos / (base ^ (2 * i / d))
            pe[pos + 1, 2 * i + 1] = sin(theta)
            pe[pos + 1, 2 * i + 2] = cos(theta)
        end
    end
    return pe
end


function apply_rope(x::Vector{Float64}, pos::Int; base::Float64=10000.0)::Vector{Float64}
    d = length(x)
    iseven(d) || throw(ArgumentError("RoPE requires an even embedding dimension"))
    out = copy(x)
    for i in 0:(d ÷ 2 - 1)
        theta = pos / (base ^ (2 * i / d))
        c = cos(theta)
        s = sin(theta)
        a = x[2 * i + 1]
        b = x[2 * i + 2]
        out[2 * i + 1] = a * c - b * s
        out[2 * i + 2] = a * s + b * c
    end
    return out
end


function dotprod(a::Vector{Float64}, b::Vector{Float64})::Float64
    return sum(a .* b)
end


function alibi_slopes(n_heads::Int)::Vector{Float64}
    n_heads > 0 || throw(ArgumentError("n_heads must be > 0"))
    return [2.0 ^ (-8.0 * (h) / n_heads) for h in 1:n_heads]
end


function alibi_bias(n_heads::Int, seq_len::Int; causal::Bool=true)
    slopes = alibi_slopes(n_heads)
    out = Vector{Matrix{Float64}}()
    for m in slopes
        bias = fill(0.0, seq_len, seq_len)
        for i in 1:seq_len
            for j in 1:seq_len
                if causal && j > i
                    bias[i, j] = -Inf
                else
                    bias[i, j] = -m * abs(i - j)
                end
            end
        end
        push!(out, bias)
    end
    return out
end


function demo_sinusoidal()
    println("=== sinusoidal positional encoding ===")
    pe = sinusoidal_pe(8, 8)
    println("first 4 positions, first 4 dims:")
    for pos in 1:4
        row_str = join([@sprintf("%+.3f", pe[pos, j]) for j in 1:4], "  ")
        @printf("  pos=%d: %s\n", pos - 1, row_str)
    end
    println()
end


function demo_rope_relative()
    println("=== RoPE: dot product depends only on relative distance ===")
    rng = MersenneTwister(0)
    d = 16
    q = randn(rng, d)
    k = randn(rng, d)
    pairs = [(3, 5), (7, 9), (100, 102), (1024, 1026)]
    @printf("%6s  %6s  %4s  %18s\n", "pos_q", "pos_k", "gap", "<q_rot, k_rot>")
    for (pq, pk) in pairs
        q_rot = apply_rope(q, pq)
        k_rot = apply_rope(k, pk)
        d_prod = dotprod(q_rot, k_rot)
        @printf("%6d  %6d  %4d  %18.6f\n", pq, pk, pk - pq, d_prod)
    end
    println("All rows with gap=2 should produce matching dot products.")
    println()
end


function demo_rope_base_scaling()
    println("=== RoPE base scaling (NTK-aware for long context) ===")
    rng = MersenneTwister(1)
    d = 8
    q = randn(rng, d)
    k = randn(rng, d)
    for base in (10000.0, 100000.0, 1_000_000.0)
        q_rot = apply_rope(q, 4096; base=base)
        k_rot = apply_rope(k, 4098; base=base)
        @printf("  base=%8d  score=%+.6f\n", Int(base), dotprod(q_rot, k_rot))
    end
    println("Larger base = slower rotation = longer context without phase wrap.")
    println()
end


function demo_alibi()
    println("=== ALiBi bias matrix ===")
    n_heads = 4
    slopes = alibi_slopes(n_heads)
    @printf("Slopes for %d heads: %s\n", n_heads,
            join([@sprintf("%.4f", s) for s in slopes], ", "))
    bias = alibi_bias(n_heads, 6; causal=false)
    println("Head 1 bias (closer tokens get smaller penalty):")
    for row in eachrow(bias[1])
        println("  " * join([@sprintf("%+6.2f", v) for v in row], "  "))
    end
    println()
end


function main()
    demo_sinusoidal()
    demo_rope_relative()
    demo_rope_base_scaling()
    demo_alibi()
    println("takeaway: RoPE encodes relative position inside the dot product;")
    println("ALiBi skips embeddings entirely. Sinusoidal is now a footnote.")
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: phases/07-transformers-deep-dive/04-positional-encoding/code/main.rs
// Positional encodings: sinusoidal, RoPE, ALiBi. Stdlib only.
// Topic: encode token position into queries, keys, or attention bias.
// References (cited in spirit, not as deps):
//   - Vaswani 2017 (sinusoidal):     https://arxiv.org/abs/1706.03762
//   - Su et al. 2021 (RoPE):         https://arxiv.org/abs/2104.09864
//   - Press et al. 2021 (ALiBi):     https://arxiv.org/abs/2108.12409
//   - candle rope impl:              https://github.com/huggingface/candle/blob/main/candle-nn/src/rotary_emb.rs
//
// Compile + run:  rustc --edition 2021 main.rs -o /tmp/pe && /tmp/pe

use std::f32::consts::PI;

// Sinusoidal positional encoding table [n, d].
fn sinusoidal_pe(n: usize, d: usize, base: f32) -> Vec<Vec<f32>> {
    let mut pe = vec![vec![0.0f32; d]; n];
    for pos in 0..n {
        for i in 0..(d / 2) {
            let theta = (pos as f32) / base.powf(2.0 * i as f32 / d as f32);
            pe[pos][2 * i] = theta.sin();
            pe[pos][2 * i + 1] = theta.cos();
        }
    }
    pe
}

// Rotate even/odd pairs of x by angle pos * theta_i. Returns a new Vec.
fn apply_rope(x: &[f32], pos: usize, base: f32) -> Vec<f32> {
    let d = x.len();
    let mut out = x.to_vec();
    for i in 0..(d / 2) {
        let theta = (pos as f32) / base.powf(2.0 * i as f32 / d as f32);
        let c = theta.cos();
        let s = theta.sin();
        let a = x[2 * i];
        let b = x[2 * i + 1];
        out[2 * i] = a * c - b * s;
        out[2 * i + 1] = a * s + b * c;
    }
    out
}

fn dot(a: &[f32], b: &[f32]) -> f32 {
    a.iter().zip(b.iter()).map(|(x, y)| x * y).sum()
}

// ALiBi slopes: 2^(-8*(h+1)/n_heads) for h in 0..n_heads.
fn alibi_slopes(n_heads: usize) -> Vec<f32> {
    (0..n_heads)
        .map(|h| 2.0f32.powf(-8.0 * (h + 1) as f32 / n_heads as f32))
        .collect()
}

// ALiBi bias matrix for each head: -slope * |i - j|, with optional causal mask.
fn alibi_bias(n_heads: usize, seq_len: usize, causal: bool) -> Vec<Vec<Vec<f32>>> {
    let slopes = alibi_slopes(n_heads);
    let mut out = Vec::with_capacity(n_heads);
    for &m in &slopes {
        let mut head = vec![vec![0.0f32; seq_len]; seq_len];
        for i in 0..seq_len {
            for j in 0..seq_len {
                head[i][j] = if causal && j > i {
                    f32::NEG_INFINITY
                } else {
                    -m * (i as i64 - j as i64).abs() as f32
                };
            }
        }
        out.push(head);
    }
    out
}

// Tiny LCG for deterministic Gaussian samples.
struct Rng { state: u64 }
impl Rng {
    fn new(seed: u64) -> Self { Rng { state: seed.wrapping_mul(0x9E37_79B9_7F4A_7C15) | 1 } }
    fn next_u32(&mut self) -> u32 {
        self.state = self.state.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        (self.state >> 33) as u32
    }
    fn uniform(&mut self) -> f32 { (self.next_u32() as f32 + 1.0) / (u32::MAX as f32 + 2.0) }
    fn gauss(&mut self) -> f32 {
        let u1 = self.uniform();
        let u2 = self.uniform();
        (-2.0 * u1.ln()).sqrt() * (2.0 * PI * u2).cos()
    }
}

fn demo_sinusoidal() {
    println!("=== sinusoidal positional encoding ===");
    let pe = sinusoidal_pe(8, 8, 10000.0);
    println!("first 4 positions, first 4 dims:");
    for pos in 0..4 {
        print!("  pos={}: ", pos);
        for i in 0..4 {
            print!(" {:+.3}", pe[pos][i]);
        }
        println!();
    }
    println!();
}

fn demo_rope_relative() {
    println!("=== RoPE: dot product depends only on relative distance ===");
    let mut rng = Rng::new(0);
    let d = 16usize;
    let q: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();
    let k: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();

    let pairs = [(3usize, 5usize), (7, 9), (100, 102), (1024, 1026)];
    println!(" {:>6}  {:>6}  {:>4}  {:>18}", "pos_q", "pos_k", "gap", "<q_rot, k_rot>");
    for (pq, pk) in pairs {
        let q_rot = apply_rope(&q, pq, 10000.0);
        let k_rot = apply_rope(&k, pk, 10000.0);
        let d_prod = dot(&q_rot, &k_rot);
        println!(" {:>6}  {:>6}  {:>4}  {:>18.6}", pq, pk, (pk as i64) - (pq as i64), d_prod);
    }
    println!("all rows with gap=2 share the same dot product.");
    println!();
}

fn demo_rope_base_scaling() {
    println!("=== RoPE base scaling (NTK-aware for long context) ===");
    let mut rng = Rng::new(1);
    let d = 8usize;
    let q: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();
    let k: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();

    for base in [10_000.0f32, 100_000.0, 1_000_000.0] {
        let q_rot = apply_rope(&q, 4096, base);
        let k_rot = apply_rope(&k, 4098, base);
        println!("  base={:>9}  score={:+.6}", base as u64, dot(&q_rot, &k_rot));
    }
    println!("larger base = slower rotation = longer context without phase wrap.");
    println!();
}

fn demo_alibi() {
    println!("=== ALiBi bias matrix ===");
    let n_heads = 4usize;
    let slopes = alibi_slopes(n_heads);
    print!("slopes for {} heads:", n_heads);
    for s in &slopes { print!(" {:.4}", s); }
    println!();
    let bias = alibi_bias(n_heads, 6, false);
    println!("head 0 bias (closer tokens get smaller penalty):");
    for row in &bias[0] {
        print!(" ");
        for v in row { print!(" {:+6.2}", v); }
        println!();
    }
    println!();
}

fn demo_rope_microbench() {
    println!("=== microbench: 50K RoPE rotations (d=128) ===");
    let mut rng = Rng::new(2);
    let d = 128usize;
    let q: Vec<f32> = (0..d).map(|_| rng.gauss()).collect();
    let start = std::time::Instant::now();
    let mut sink = 0.0f32;
    for pos in 0..50_000usize {
        let r = apply_rope(&q, pos, 10_000.0);
        sink += r[0];
    }
    let elapsed = start.elapsed();
    println!("50K rotations in {:.2}ms ({:.0}/sec)  sink={:.4}",
        elapsed.as_secs_f64() * 1000.0,
        50_000.0 / elapsed.as_secs_f64(),
        sink,
    );
}

fn main() {
    demo_sinusoidal();
    demo_rope_relative();
    demo_rope_base_scaling();
    demo_alibi();
    demo_rope_microbench();
    println!();
    println!("takeaway: RoPE encodes relative position in the dot product itself.");
    println!("ALiBi skips embeddings entirely. sinusoidal is mostly historical now.");
}
